### PR TITLE
ipc: handler and helper partitioning

### PR DESCRIPTION
**Pull request by lgirdwood**
Continued partitioning and cleanup of IPC infrastructure to support more MAJOR ABIs

### CODE REVIEW

**Comment by marc-hb** [comment]
https://sof-ci.01.org/sofpr/PR4029/build8677/devicetest failed because TGLU_SKU0A32_SDCA was added to PR testing before being ready, sorry for the inconvenience. I removed it.

**Comment by marc-hb** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
Looks like UTs need to link with the IPC helper

**Comment by lgirdwood** [comment]
@zrombel fyi - looks like CI pending for a while.

File: src/audio/pipeline/pipeline-graph.c [hunk 1]
```diff
@@ -123,13 +122,12 @@ struct pipeline *pipeline_new(struct sof_ipc_pipe_new *pipe_desc,
 
 	/* init pipeline */
 	p->sched_comp = cd;
+	p->comp_id = comp_id;
+	p->pipeline_id = pipeline_id;
```
Comment (lyakh) [inline src/audio/pipeline/pipeline-graph.c:126]: no need to set `.priority`?

File: src/audio/buffer.c [hunk 2]
```diff
@@ -70,35 +70,6 @@ struct comp_buffer *buffer_alloc(uint32_t size, uint32_t caps, uint32_t align)
...
-
-		memcpy_s(&buffer->tctx, sizeof(struct tr_ctx),
-			 &buffer_tr, sizeof(struct tr_ctx));
-
-		dcache_writeback_invalidate_region(buffer, sizeof(*buffer));
-	}
-
-	return buffer;
-}
-
```
Comment (keyonjie) [inline src/audio/buffer.c:101]: do we plan to use comp_buffer as a component in ipc4?  if no, maybe better to move the whole file to ABI specific.

File: src/audio/pipeline/pipeline-schedule.c [hunk 3]
```diff
@@ -94,6 +94,20 @@ static struct task *pipeline_task_init(struct pipeline *p, uint32_t type,
...
+			     uint32_t core, uint32_t period,
+			     uint32_t period_mips, uint32_t frames_per_sched,
+			     uint32_t time_domain)
+{
+	p->sched_id = sched_id;
+	p->core = core;
+	p->period = period;
+	p->period_mips = period_mips;
+	p->frames_per_sched = frames_per_sched;
+	p->time_domain = time_domain;
```
Comment (lyakh) [inline src/audio/pipeline/pipeline-schedule.c:107]: interesting, looks like some of these parameters weren't used at all until now?

File: src/audio/pipeline/pipeline-graph.c (hunk 1 above)
Comment (lgirdwood) [reply src/audio/pipeline/pipeline-graph.c:127]: your right, and it's not been used today. Lets keep it as we will use it in the future. Now fixed.

File: src/audio/pipeline/pipeline-schedule.c (hunk 3 above)
Comment (lgirdwood) [reply src/audio/pipeline/pipeline-schedule.c:107]: They were, but the pipeline contained the IPC specific structure and it copied it in bulk.

**Review by keyonjie — COMMENTED** [review]
The pipeline and cmoka change looks somewhat invasive, better to run more test with it IMHO.

File: src/audio/buffer.c (hunk 2 above)
Comment (lgirdwood) [reply src/audio/buffer.c:101]: This will be incrementally updated.

**Review by lyakh — APPROVED** [review]

File: src/ipc/handler-ipc3.c [hunk 4]
```diff
@@ -62,6 +62,12 @@
 #include <stddef.h>
 #include <stdint.h>
 
+#if CONFIG_CAVS && CAVS_VERSION >= CAVS_VERSION_1_8
+#include <ipc/header-intel-cavs.h>
+#include <cavs/drivers/sideband-ipc.h>
+#define CAVS_IPC_TYPE_S(x)		((x) & CAVS_IPC_TYPE_MASK)
+#endif
+
```
[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/ipc/handler-ipc3.c:70]: Now fixed, in fact the IPC includes could also be removed from this file (and have been).

**Comment by lgirdwood** [comment]
> The pipeline and cmoka change looks somewhat invasive, better to run more test with it IMHO.

Been locally tested, this is all mechanical code moves so the pipeline runtime logic is the same. i.e. we are no longer using the IPC structure embedded within pipeline, but using the member directly now.

The cmocka changes actually increase the code test coverage so the fact they are passing on more FW code is an improvement.

**Review by keyonjie — APPROVED** [review]

**Comment by lgirdwood** [comment]
@zrombel tests were passing before I changed commit message and now they are not visible. Can you confirm ?

**Comment by lgirdwood** [comment]
Jenkins showing known kernel PM and module unload issue.

**Pull request merged**